From 250505bc7361ea1952d519ce77b83212021c5fed Mon Sep 17 00:00:00 2001
From: Hariharan Basuthkar <quic_hbasuthk@quicinc.com>
Date: Tue, 16 Jul 2024 22:08:50 +0530
Subject: [PATCH] hostapd: Add single wiphy support to AFC vendor QCA commands
 and events

Add single wiphy support to the AFC vendor QCA commands and events
mentioned below:

  1) QCA_NL80211_VENDOR_SUBCMD_SET_6GHZ_POWER_MODE
  2) QCA_NL80211_VENDOR_SUBCMD_POWER_MODE_CHANGE_COMPLETED and
  3) QCA_NL80211_VENDOR_SUBCMD_AFC_EVENT

Patch dependency: r32-001-hostapd-Add-support-to-receive-AFC-power-update-comp.patch
                  r28-hostapd-Add-hostapd-cli-support-to-switch-6GHZ-power.patch

Tag: 6G_REGULATORY

Signed-off-by: Hariharan Basuthkar <quic_hbasuthk@quicinc.com>
---
 src/drivers/driver_nl80211.c       |  2 +-
 src/drivers/driver_nl80211_event.c | 20 +++++++++++---------
 2 files changed, 12 insertions(+), 10 deletions(-)

--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -11871,7 +11871,7 @@ static int nl80211_set_6ghz_pwr_mode(voi
 	if (drv->nlmode != NL80211_IFTYPE_AP)
 		return -EOPNOTSUPP;
 
-	if (!(msg = nl80211_drv_msg(drv, 0, NL80211_CMD_VENDOR)) ||
+	if (!(msg = nl80211_bss_msg(bss, 0, NL80211_CMD_VENDOR)) ||
 	    nla_put_u32(msg, NL80211_ATTR_VENDOR_ID, OUI_QCA) ||
 	    nla_put_u32(msg, NL80211_ATTR_VENDOR_SUBCMD,
 			QCA_NL80211_VENDOR_SUBCMD_SET_6GHZ_POWER_MODE)) {
--- a/src/drivers/driver_nl80211_event.c
+++ b/src/drivers/driver_nl80211_event.c
@@ -3482,7 +3482,7 @@ static void qca_nl80211_pasn_auth(struct
 #endif /* CONFIG_DRIVER_NL80211_QCA */
 
 static void
-qca_nl80211_6ghz_pwr_mode_change_completed(struct wpa_driver_nl80211_data *drv,
+qca_nl80211_6ghz_pwr_mode_change_completed(struct i802_bss *bss,
 					   u8 *data, size_t len)
 {
 	union wpa_event_data event;
@@ -3510,7 +3510,7 @@ qca_nl80211_6ghz_pwr_mode_change_complet
 	wpa_printf(MSG_INFO, "nl80211: 6GHZ power mode changed %d",
 		   event.ap_6ghz_pwr_mode_event.pwr_mode);
 
-	wpa_supplicant_event(drv->ctx, EVENT_6GHZ_POWER_MODE_NOTIFY, &event);
+	wpa_supplicant_event(bss->ctx, EVENT_6GHZ_POWER_MODE_NOTIFY, &event);
 }
 
 
@@ -3724,7 +3724,7 @@ static int copy_afc_chan_obj(struct nlat
 
 #define NUM_6GHZ_OPCLASS 7
 static int
-qca_nl80211_afc_power_update_completed(struct wpa_driver_nl80211_data *drv,
+qca_nl80211_afc_power_update_completed(struct i802_bss *bss,
 				       u8 *data, size_t len)
 {
 	union wpa_event_data event;
@@ -3867,7 +3867,7 @@ qca_nl80211_afc_power_update_completed(s
 		afc_rsp->afc_chan_info = afc_chan_info;
 	}
 
-	wpa_supplicant_event(drv->ctx, EVENT_AFC_POWER_UPDATE_COMPLETE_NOTIFY,
+	wpa_supplicant_event(bss->ctx, EVENT_AFC_POWER_UPDATE_COMPLETE_NOTIFY,
 			     &event);
 	os_free(afc_freq_info);
 	os_free(afc_chan_info);
@@ -3875,9 +3875,11 @@ qca_nl80211_afc_power_update_completed(s
 	return 0;
 }
 
-static void nl80211_vendor_event_qca(struct wpa_driver_nl80211_data *drv,
+static void nl80211_vendor_event_qca(struct i802_bss *bss,
 				     u32 subcmd, u8 *data, size_t len)
 {
+	struct wpa_driver_nl80211_data *drv = bss->drv;
+
 	switch (subcmd) {
 	case QCA_NL80211_VENDOR_SUBCMD_TEST:
 		wpa_hexdump(MSG_DEBUG, "nl80211: QCA test event", data, len);
@@ -3921,10 +3923,10 @@ static void nl80211_vendor_event_qca(str
 		break;
 #endif /* CONFIG_DRIVER_NL80211_QCA */
 	case QCA_NL80211_VENDOR_SUBCMD_POWER_MODE_CHANGE_COMPLETED:
-		qca_nl80211_6ghz_pwr_mode_change_completed(drv, data, len);
+		qca_nl80211_6ghz_pwr_mode_change_completed(bss, data, len);
 		break;
 	case QCA_NL80211_VENDOR_SUBCMD_AFC_EVENT:
-		qca_nl80211_afc_power_update_completed(drv, data, len);
+		qca_nl80211_afc_power_update_completed(bss, data, len);
 		break;
 	default:
 		wpa_printf(MSG_DEBUG,
@@ -4016,9 +4018,10 @@ static void nl80211_vendor_event_brcm(st
 #endif /* CONFIG_DRIVER_NL80211_BRCM */
 
 
-static void nl80211_vendor_event(struct wpa_driver_nl80211_data *drv,
+static void nl80211_vendor_event(struct i802_bss *bss,
 				 struct nlattr **tb)
 {
+	struct wpa_driver_nl80211_data *drv = bss->drv;
 	u32 vendor_id, subcmd, wiphy = 0;
 	int wiphy_idx;
 	u8 *data = NULL;
@@ -4064,7 +4067,7 @@ static void nl80211_vendor_event(struct
 
 	switch (vendor_id) {
 	case OUI_QCA:
-		nl80211_vendor_event_qca(drv, subcmd, data, len);
+		nl80211_vendor_event_qca(bss, subcmd, data, len);
 		break;
 #ifdef CONFIG_DRIVER_NL80211_BRCM
 	case OUI_BRCM:
@@ -4823,7 +4826,7 @@ static void do_process_drv_event(struct
 		nl80211_stop_ap(bss, tb);
 		break;
 	case NL80211_CMD_VENDOR:
-		nl80211_vendor_event(drv, tb);
+		nl80211_vendor_event(bss, tb);
 		break;
 	case NL80211_CMD_NEW_PEER_CANDIDATE:
 		nl80211_new_peer_candidate(drv, tb);
